### PR TITLE
set content-type as image/xxx to avoid pontential XSS risk

### PR DESCRIPTION
**Pull request by rozhao2**

### CODE REVIEW

File: thumberd/thumberd.go
Comment (rozhao2) [inline thumberd/thumberd.go:586]: Add content type for supported image types

Comment (rozhao2) [inline thumberd/thumberd.go:587]: Add content type for supported image types

File: thumberd/thumberd.go
Comment (rozhao2) [inline thumberd/thumberd.go:540]: return detected image type together

Comment (rozhao2) [inline thumberd/thumberd.go:539]: return detected image type together

File: thumberd/thumberd.go
Comment (rozhao2) [inline thumberd/thumberd.go:509]: If request doesn't assign target format, use the detected image content type instead of Content-Type from request header.

So it will make sure response Content-Type is image MIME-TYPE, exclude XSS possibility.

Comment (rozhao2) [inline thumberd/thumberd.go:508]: If request doesn't assign target format, use the detected image content type instead of Content-Type from request header.

So it will make sure response Content-Type is image MIME-TYPE, exclude XSS possibility.

File: thumberd/thumberd.go
Comment (shadowstriker9) [inline thumberd/thumberd.go]: Maybe error code 400 (Bad Request) makes more sense in this case?

**Review by shadowstriker9 — COMMENTED** [review]
If possible could you add some unit testing for your new logic?

File: thumberd/thumberd.go
Comment (rozhao2) [reply thumberd/thumberd.go]: Sure return bad request instead, thanks

**Comment by rozhao2** [comment]
> If possible could you add some unit testing for your new logic?

Ya, agree, actually I tried but failed, thumberd_test.go file was commited 6 years ago and never changed, go test reports type error

```
# _/Users/rong/smartnews/projects/yoya-thumber/thumberd [_/Users/rong/smartnews/projects/yoya-thumber/thumberd.test]
./thumberd_test.go:12:44: cannot convert thumbServer (value of type func(w http.ResponseWriter, r *http.Request, sem chan int)) to type http.HandlerFunc
./thumberd_test.go:30:44: cannot convert thumbServer (value of type func(w http.ResponseWriter, r *http.Request, sem chan int)) to type http.HandlerFunc
./thumberd_test.go:50:44: cannot convert thumbServer (value of type func(w http.ResponseWriter, r *http.Request, sem chan int)) to type http.HandlerFunc
./thumberd_test.go:71:44: cannot convert thumbServer (value of type func(w http.ResponseWriter, r *http.Request, sem chan int)) to type http.HandlerFunc
```

It is caused by http.HandlerFunc expecting a function with (w http.ResponseWriter, r *http.Request), but thumbServer is defined as `thumbServer(w http.ResponseWriter, r *http.Request, sem chan int)`.

This function was defined 6 years ago too, if there was no API interface change of golang http module, it means this repo's unit test never run successfully, lol... 

Then I tried to fix it with a wrapper, but still failed with error `flag provided but not defined: -test.paniconexit0`.

It was caused by `flag.Parse()` was called in init() function, in simple words, we may need extra time on fixing the unit test...

At the same time, I deployed new change on staging instance and tested postive/negative cases, for the test URL [https://lcamtuf.coredump.cx/squirrel/](http://127.0.0.1/?url=https://lcamtuf.coredump.cx/squirrel/), we can see the response content type is image/jpeg instead of text/html which will prevent client's browser executing malicious scripts as my understanding.

```
# curl -v http://127.0.0.1/?url=https://lcamtuf.coredump.cx/squirrel/                                                                                                                                                                                    
*   Trying 127.0.0.1...                                                                                                                                                                                                                                                         
* TCP_NODELAY set                                                                                                                                                                                                                                                               
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)                
> GET /?url=https://lcamtuf.coredump.cx/squirrel/ HTTP/1.1          
> Host: 127.0.0.1                                                                                                                                                                                                                                                               
> User-Agent: curl/7.53.1                                                                                                               
> Accept: */*                                 
>                                                                   
< HTTP/1.1 200 OK                                                                                                                       
< Server: nginx                                  
< Date: Wed, 27 Apr 2022 12:59:28 GMT                                                                                                   
< Content-Type: image/jpeg                                                                                                                                                                                                                                                      
< Content-Length: 119788                                                                                                                
< Connection: keep-alive                                                                                                                
< Last-Modified: Wed, 27 Apr 2022 12:59:27 GMT                                                                                          
<                                                                                                                                       
JFIF,,r<html><body><style>body {
```

Thanks.

**Review by shadowstriker9 — APPROVED** [review]
I'm not an expert but LGTM! Let's wait for @sn-nyamamoto to review from a security stand point just to be safe

File: thumberd/thumberd.go
Comment (lukas-sn) [inline thumberd/thumberd.go:484]: is this necessary in prod code? (sorry I'm not very familiar with golang)

Comment (lukas-sn) [reply thumberd/thumberd.go:484]: and I've noticed this is not written by you but rather from the old code

**Review by lukas-sn — APPROVED** [review]

**Review by shadowstriker9 — APPROVED** [review]
I already [reviewed](https://github.com/smartnews/yoya-thumber/pull/43#pullrequestreview-961333242) I'm not sure why it's not showing as approved from me... 🤔 
image

File: thumberd/thumberd.go
Comment (honea-sn) [inline thumberd/thumberd.go]: Can we change the error message to:
"Invalid data retrieved"

File: thumberd/thumberd.go
Comment (honea-sn) [inline thumberd/thumberd.go]: Can we change the error message to:
"Invalid data retrieved"

File: thumberd/thumberd.go
Comment (rozhao2) [reply thumberd/thumberd.go:484]: Thank you for checking, thumbnail service will run in docker, all stdout output will be redirected to log file, so we feel it is a print to stdout, it will become log actually. 

It is more about container deployment than golang, I think there is argument on how to log in container environment... but let us keep old style at present.

```
stdout_logfile=/data/log/app/thumberd.log
```

File: thumberd/thumberd.go
Comment (rozhao2) [reply thumberd/thumberd.go]: Sure, updated.

File: thumberd/thumberd.go
Comment (rozhao2) [reply thumberd/thumberd.go]: Sure, updated.

**Pull request merged**
